File: 0055-jump-game/0055-jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        if (nums.length == 1) return true;
        
        int r = nums.length - 1;
        int l = nums.length - 2;

        while (l >= 0) {
            if ((r - l) - nums[l] <= 0) {
                r = l;
                if (r == 0) return true;
                l--;
            } else {
                l--;
            }
        }
        return false;
    }
}

File: 0055-jump-game/0055-jump-game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int currJump = 0;

        for (int i = 0; i < nums.size(); i++) {
            currJump = std::max(currJump, nums[i]);
            
            if (currJump + i >= nums.size() - 1) return true;
            if (currJump <= 0) return false;

            currJump--;
        }

        return false;
    }
};